Introduce `@extern(c)` to declare C function without Clang module

kateinoigakukun opened the pull request:
## Motivation

Currently the only **proper** way to reference a C function from Swift is by importing it through Clang module.
However there are a number of places using the incorrect `@_silgen_name` way to reference C function. (The attribute expects Swift calling convention and it causes calling convention mismatch.)
These incorrect references happened because defining a new Clang module requires non-trivial work mostly around the build system. (Also there is a very complex issue, which is not easy to solve with Clang module approach https://github.com/apple/swift/pull/67853#issuecomment-1760710859)

So this PR adds a new variant of `@extern` for C function declaration. Also this unlocks a way to declare a function with C calling convention imported through Wasm’s import mechanism by mixing with [`@extern(wasm)`](https://github.com/apple/swift/pull/69107)

The attribute is now gated by `-enable-experimental-feature Extern`

## Description

The new attribute unlocks declaring a C function in Swift source code. The declared function interfaces are limited to be C-compatible, so no async, no throws, and no ref counted types are allowed. (We might be able to unlock the use of ref counted type in theory but it introduces a lot of complexities)

Examples:

```swift
@extern(c)
func malloc(size: Int) -> UnsafeMutableRawPointer!

@extern(c, "swift_demangle")
public func _stdlib_demangleImpl(
  mangledName: UnsafePointer<CChar>?,
  mangledNameLength: UInt,
  outputBuffer: UnsafeMutablePointer<CChar>?,
  outputBufferSize: UnsafeMutablePointer<UInt>?,
  flags: UInt32
) -> UnsafeMutablePointer<CChar>?

@extern(wasm, module: "wasi_snapshot_preview1", name: "random_get")
@extern(c) // NOTE: Just for calling it with C calling convention
func wasi_random_get(
  buffer: UnsafeMutablePointer<UInt8>,
  count: UInt32
) -> Int32
```

Resolves rdar://115802180 and unblock https://github.com/apple/swift/pull/67853

<details>
<summary>Alternative consideration: Allow @_cdecl without body</summary>

We initially planned this approach, but didn’t take it for the following reason

* The current semantics of `@_cdecl` is “expose the Swift declaration to C/Objective-C world” and it allows having Objective-C types in its interface.
    * Thus it requires a native-to-foreign thunk to be emitted within the module, which defines the `@_cdecl` function.
* The new mode (without body) would add a semantics: “if no function body is defined, reference a C function corresponding to the Swift function declaration”.
    * This requires a foreign-to-native thunk to be emitted within the module, which *declare* the `@_cdecl` function.
* The problem here is we have to determine if a function with @_cdecl requires foreign-to-native thunk based on where function declarations come from because a @_cdecl function without body declared within a compiling module and imported @_cdecl function both do not have body.

</details>

kateinoigakukun committed:
[c-interop] Add @_extern(c) attribute to declare C function in Swift

kateinoigakukun committed:
[c-interop] Use llvm::Optional instead of std::optional for now

The Attr.h is shared with SwiftCompilerSources through C++ interop and
C++ interop somehow crashes with libc++'s std::optional. So use legacy
llvm::Optional for now.

kateinoigakukun commented:
@swift-ci Please smoke test

kateinoigakukun committed:
[c-interop][test] Disable availablity checks to test async interface

kateinoigakukun committed:
[c-interop][test] Re-write call abi checks with C struct in more flexible way

To fix windows-x86_64 case, which doesn't have byval.

kateinoigakukun commented:
@swift-ci Please smoke test

kateinoigakukun commented:
@swift-ci Please smoke test Apple Silicon

kateinoigakukun commented:
@swift-ci Please test Apple Silicon

al45tair commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2065,11 +2065,83 @@ void AttributeChecker::visitExposeAttr(ExposeAttr *attr) {
+  if (D->getAttrs().hasAttribute<CDeclAttr>())
+    diagnose(attr->getLocation(), diag::extern_only_non_other_attr, "@_cdecl");
+
+  if (!isCCompatibleFuncDecl(FD)) {
```
Does this not need to be inside the `ExternKind::C` condition? What happens when this visits an `ExternKind::Wasm` `ExternAttr`? Is the set of Wasm compatible function declarations equal to the set of C compatible function declarations? If so, maybe we should rename `isCCompatibleFuncDecl()`? If not, does the Wasm code not need to do its own check here?

kateinoigakukun replied:
Oops, this check should be only for `@_extern(c)`. `@_extern(wasm)` should be able to reference any signature and have the same representation as usual Swift function. Thank you for catching this 🙏

al45tair commented on test/attr/attr_extern.swift:
```diff
@@ -37,5 +37,71 @@ func f11Scope() {
+
+@_extern(c)
+func genericFuncC<T>(_: T) // expected-error {{'T' cannot be represented in C}}
+
```
What happens with these validation cases where you have e.g.
```
@_extern(c)
@_extern(wasm, module: "foo", name: "bar")
func nonCReturnTypesMixed() -> NonC
```
or even just
```
@_extern(wasm, module: "foo", name: "bar")
func nonCReturnTypesWasm() -> NonC
```
Are there return types Wasm doesn't support? Do we get an error message about the type not being representable in C in that case?

kateinoigakukun replied:
Added test cases

kateinoigakukun committed:
[c-interop] Check interface C compatibility only for `@_extern(c)`

`@_extern(wasm)` has no limitation on the function interface, so the
check should be applied only for `@_extern(c)`.

kateinoigakukun committed:
[c-interop][docs] Add @_extern(c) attribute explanation

kateinoigakukun commented:
@swift-ci Please smoke test

al45tair commented on docs/ReferenceGuides/UnderscoredAttributes.md:
```diff
@@ -449,6 +449,20 @@ through WebAssembly's import interface.
+
+Indicates that a particular declaration should refer to a
+C declaration with the given name. If the optional "cName"
+string is not specified, the mangled Swift name will be used.
```
This seems _surprising_. I would expect
```
@_extern(c)
func malloc(_ size: UInt) -> UnsafeRawPointer?
```
to call a function called `malloc`, not `$s4tst26mallocySVSgSuF`.

kateinoigakukun replied:
I used the naming strategy to keep the same behavior with `@silgen_name`. But on second thought, using names without Swift mangling scheme seems more natural since no one defines C function with a mangled Swift name 👍

al45tair commented on docs/ReferenceGuides/UnderscoredAttributes.md:
:-) I meant to ask you to update this, but forgot when I last commented on the PR.

al45tair commented on docs/ReferenceGuides/UnderscoredAttributes.md:
```diff
@@ -449,6 +449,20 @@ through WebAssembly's import interface.
+C declaration with the given name. If the optional "cName"
+string is not specified, the mangled Swift name will be used.
+
+Similar to `@_cdecl`, but this attribute is used to reference
+C declarations from Swift, while `@_cdecl` is used to be referenced
+from C.
```
The latter part of this sentence is difficult to parse.
```suggestion
Similar to `@_cdecl`, but this attribute is used to reference C declarations from Swift, while `@_cdecl` is used to define Swift functions that can be referenced from C.
```

al45tair commented on test/IRGen/extern_c.swift:
```diff
@@ -0,0 +1,16 @@
+// CHECK: declare void @explicit_extern_c()
+@_extern(c, "explicit_extern_c") func explicit_extern_c()
+
+// CHECK: declare void @"$s8extern_c09implicit_A2_cyyF"()
+@_extern(c) func implicit_extern_c()
```
See my other remark. I think this should call a function `implicit_extern_c()`. I think people would be very surprised to find that it was mangled with Swift's name mangling.

al45tair replied:
(There are other similar examples in the tests.)

kateinoigakukun committed:
[c-interop] Use Swift base name for @_extern(c) without explicit name

kateinoigakukun commented:
@swift-ci Please test

al45tair commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2065,11 +2065,85 @@ void AttributeChecker::visitExposeAttr(ExposeAttr *attr) {
+    }
+  }
+
+  // @_cdecl cannot be mixed with @_extern since @_cdecl is for definitions
+  if (D->getAttrs().hasAttribute<CDeclAttr>())
+    diagnose(attr->getLocation(), diag::extern_only_non_other_attr, "@_cdecl");
```
Should we also check here for `@_silgen_name()`? Either sense of `@_silgen_name()` would appear to clash with `@_extern()`.

al45tair commented on test/IRGen/extern_c.swift:
```diff
@@ -0,0 +1,37 @@
+@_extern(c) func implicit_extern_c()
+
+// CHECK: declare i32 @"+"({{.*}})
+@_extern(c) func +(a: c_struct, b: c_struct) -> c_struct
```
I guess we don't check that the name is a valid C symbol name? Maybe we should? You wouldn't be allowed to declare a function called `+` in C.

kateinoigakukun replied:
Yeah, good catching. "Valid" C symbol is actually implementation-defined thing in the C standard spec considering Universal Character Names, as far as I understand. I took a look in Clang to find a way to check if a string is a valid C identifier in Clang, but I couldn't find it. 
(there is `isValidAsciiIdentifier` but it does not support UCNs, and it seems any identifiers that can be lex'ed are treated as valid)

So there are two options here. 1. Accept only ASCII identifiers using [`clang::isValidAsciiIdentifier`](https://github.com/llvm/llvm-project/blob/761c9dd92789b744607dc9c8c5071fef340fd86f/clang/include/clang/Basic/CharInfo.h#L224-L238), 2. Accept any identifiers that are valid as Swift function name. 

Option 1 can reject identifiers including UCNs but can reject apparently invalid C identifiers. Option 2 accepts all valid C identifiers, but can accept invalid identifiers too.

I think Option 1 is conservative, but better approach to be practical and we can incrementally unlock UCNs things later. WDYT?

al45tair replied:
Agreed, I would go with option 1 for now. UCNs and unicode in identifiers more generally are not particularly widely used in C, and I think they're especially unlikely in situations where we'd like to use `@_extern(c)` (because that's mainly going to be used to call functions in the runtime and/or legacy C code, much of which predates Unicode identifier support anyway).

al45tair replied:
What we _could_ do is make it so that using a potentially invalid name is just a warning, rather than an error per se. That way people get told that something might go wrong, and if they feel they know better then they can carry on anyway? I'm not sure how I feel about that, but it's worth thinking about.

kateinoigakukun replied:
Considering this feature is for-expert, I think it makes sense to make it just a warning not to block anything that *might* be valid.

al45tair replied:
Maybe we only want the warning if the name is implicit. For explicit names, as you say, this is an "expert" feature, so if I _really_, _really_ want to call a function whose symbol is "+" or even "🥸", maybe it should just trust me at that point.

kateinoigakukun replied:
Ah, it seems the right way! Thanks!

al45tair commented on test/Interop/WebAssembly/extern-wasm-cdecls.swift:
```diff
@@ -0,0 +1,36 @@
+// RUN: %empty-directory(%t)
+// RUN: %target-swift-frontend %s -emit-ir -module-name Extern | %FileCheck %s
+
+// CHECK: declare void @"$s6Extern7import1yyF"() [[EA1:#[0-9]+]]
+@_extern(c)
+@_extern(wasm, module: "m0", name: "import1")
+func import1()
+
+// CHECK: declare i32 @"$s6Extern20import2WithReturnInts5Int32VyF"() [[EA2:#[0-9]+]]
+@_extern(c)
+@_extern(wasm, module: "m0", name: "import2")
+func import2WithReturnInt() -> Int32
+
+// CHECK: declare void @"$s6Extern16import3TakingIntyys5Int32VF"(i32) [[EA3:#[0-9]+]]
+@_extern(c)
+@_extern(wasm, module: "m0", name: "import3")
+func import3TakingInt(_: Int32)
```
These look like they'll fail as a result of the change of naming (sorry).

kateinoigakukun replied:
Oops, I didn't checked Interop tests 😅

al45tair commented on test/attr/attr_extern.swift:
I think we want some tests for
```
@_silgen_name("...")
@_extern(c)
func foo(...) -> Bar
```
and so on as well.

al45tair reviewed (commented):
Looking good. Still a few niggles I think.

kateinoigakukun committed:
[c-interop] Reject @_silgen_name with @_extern

To avoid ambiguity at SIL level name since both tries to change the SIL
level name.

MaxDesiatov commented:
@swift-ci test

kateinoigakukun committed:
[c-interop] Ban non-ASCII identifiers in @_extern(c) conservatively

kateinoigakukun commented:
@swift-ci Please test

kateinoigakukun committed:
[c-interop] Relax C name validation as warning

And allow expert users to use potentially invalid symbols by specifying
it explicitly.

kateinoigakukun commented:
@swift-ci Please build toolchain

al45tair reviewed (approved):
I think this is great :-)

It'd be nice to have someone else sign off on it as well though. Don't know who's the best person to look at it; @lorentey seems like a good choice because he's likely to actually use it. @mikeash might also appreciate it and I don't think is on the reviewers list.

kateinoigakukun committed:
[c-interop][test] Cover default argument values with @_extern

kateinoigakukun committed:
[c-interop] Gate @_extern behind experimental feature flag

kateinoigakukun committed:
[c-interop] Rename @_extern to @extern

Now the feature is gated by experimental feature flag.
It's not shipped in any language release, so this rename should be fine.

kateinoigakukun commented:
@swift-ci Please smoke test

kateinoigakukun commented:
@swift-ci Please smoke test

MaxDesiatov commented:
@swift-ci build toolchain

kateinoigakukun commented:
@swift-ci Please smoke test

jckarter commented:
> The new mode (without body) would add a semantics: “if no function body is defined, reference a C function corresponding to the Swift function declaration”.
> This requires a foreign-to-native thunk to be emitted within the module, which declare the @_cdecl function

Note that the foreign-to-native thunk is still necessary in general. Currently this crashes the compiler:

```
@_extern(c)
func foo(x: Int)

func bar(_: (Int) -> Void) {}

func bas() {
    bar(foo)
}
```

It would be cool if we treated these like we do functions imported from Clang and generate the foreign-to-native thunk as a shared thunk on demand.
